app/packs/src/apps/mydb/elements/details/reactions/variationsTab/ReactionVariations.js

```diff
@@ -0,0 +1,231 @@
+
+function MaterialHeader({ material }) {
+  return (
+    <OverlayTrigger placement="bottom" overlay={iupacNameTooltip(material)}>
+      <div>
+        {material.external_label || material.short_label || material.id.toString()}
+      </div>
+    </OverlayTrigger>
+  );
+}
+
+function RowToolsCellRenderer({ data, copyRow, removeRow }) {
+  return (
+    <ButtonGroup>
+      <Button onClick={() => copyRow(data)}><i className="fa fa-copy" /></Button>
+      <Button onClick={() => removeRow(data)}><i className="fa fa-trash" /></Button>
+    </ButtonGroup>
+  );
+}
+
+function CellRenderer({ value: cellData }) {
+  const { value = '', unit = 'None', aux = {} } = cellData ?? {};
+  const cellContent = `${Number(value) ? Number(value).toFixed(3) : 'NaN'} [${unit}]`;
+
+  let overlayContent = aux.coefficient ? `Coeff: ${aux.coefficient}` : '';
+  overlayContent += aux.isReference ? '; Ref' : '';
+  overlayContent += aux.yield ? `; Yield: ${aux.yield}%` : '';
+  if (!overlayContent) {
+    return cellContent;
+  }
+
+  const overlay = (
+    <Tooltip>
+      {overlayContent}
+    </Tooltip>
+  );
+  return (
+    <OverlayTrigger placement="bottom" overlay={overlay}>
+      <div>
+        {cellContent}
+      </div>
+    </OverlayTrigger>
+  );
+}
+
+const CellEditor = forwardRef((props, ref) => {
+  const { value = '', unit = 'None', aux = {} } = props.value ?? {};
+  const [editedValue, setEditedValue] = useState(value);
+  const refInput = useRef(null);
+
+  useEffect(() => {
+    // focus on the input
+    refInput.current.focus();
+  }, []);
+
+  useImperativeHandle(ref, () => ({
+    getValue() {
+      // final value to send to the grid, on completion of editing
+      return { value: editedValue, unit, aux };
+    },
+
+    isCancelAfterEnd() {
+      // validate edit here: return true to declare edit invalid and keep previous value
+      return false;
+    }
+  }));
+
+  return (
+    <input
+      type="number"
+      ref={refInput}
+      value={editedValue}
+      onChange={(event) => setEditedValue(event.target.value)}
+      style={{ width: '100%' }}
+      disabled={aux.isReference}
+    />
+  );
+});
+
+export default function ReactionVariations({ reaction, onEditVariations }) {
+  const gridRef = useRef();
+
+  const [materialUnit, setMaterialUnit] = useState('Equiv');
+
+  function addRow() {
+    const newRow = createVariationsRow(reaction, uuidv4(), materialUnit);
+    onEditVariations(
+      [...reaction.variations, newRow]
+    );
+  }
+
+  function copyRow(data) {
+    const copiedRow = _.cloneDeep(data);
+    copiedRow.id = uuidv4();
+    onEditVariations(
+      [...reaction.variations, copiedRow]
+    );
+  }
+
+  function removeRow(data) {
+    onEditVariations(reaction.variations.filter((row) => row.id !== data.id));
+  }
+
+  const updateRow = useCallback(({ data: oldRow, colDef, newValue }) => {
+    const { field } = colDef;
+    const updatedRow = { ...oldRow };
+    _.set(updatedRow, field, newValue);
```

Comment:
Was macht dieser Befehl? (Hintergrund der Frage: kann man den durch irgendwas ersetzen was kein lodash benötigt. Diese Lib ist schon steinalt und inzwischen eigentlich meines Wissens nach unnötig

Reply:
Mit `_set()` mutiere ich ein genestetes Attribut des `updatedRow` Objekts.
D.h., `field` kann `foo.bar.baz` sein. Hab nach relativ langer Suche keine equivalente Loesung in reinem JS gefunden. Mit der [Spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) z.B., kann man keine genesteten keys verwenden: `{ ...oldRow, foo.bar.baz: 'value' }` bzw. `{ ...oldRow, 'foo.bar.baz': 'value' }` funktioniert nicht.

Ich verwende lodash, da die Abhaengigkeit schon in der Applikation war. Urspruenglich brauchte ich nur `cloneDeep()` aus lodash, siehe z.B., `removeObsoleteMaterialsFromVariations`, `addMissingMaterialsToVariations` in `ReactionVariationsUtils`.
Da ich lodash eh schon benutze, erschien es mir weniger komplex `_set()` zu verwenden als eine hausgemachte Funktion fuer das setzen genesteter keys zu schreiben (ala https://stackoverflow.com/questions/18936915/dynamically-set-property-of-nested-object).